rust/lance-encoding/src/encodings/physical/bitpack.rs

```diff
@@ -0,0 +1,650 @@
+    let num_bits = arr.data_type().byte_width() as u64 * 8 - max.leading_zeros() as u64;
+
+    // we can't bitpack into 0 bits, so the minimum is 1
+    num_bits.max(1)
+}
+
+#[derive(Debug)]
+pub struct BitpackingBufferEncoder {
+    num_bits: u64,
+}
+
+impl BitpackingBufferEncoder {
+    pub fn new(num_bits: u64) -> Self {
+        Self { num_bits }
+    }
+}
+
+impl BufferEncoder for BitpackingBufferEncoder {
+    fn encode(&self, arrays: &[ArrayRef]) -> Result<(EncodedBuffer, EncodedBufferMeta)> {
+        // calculate the total number of bytes we need to allocate for the destination.
+        // this will be the number of items in the source array times the number of bits.
+        // let count_items = count_items_to_pack(arrays);
+        let count_items = arrays.iter().map(|arr| arr.len()).sum::<usize>();
+        let dst_bytes_total = count_items * ceil(self.num_bits as usize, 8);
+
+        let mut dst_buffer = vec![0u8; dst_bytes_total];
+        let mut dst_idx = 0;
+        let mut dst_offset = 0;
+        for arr in arrays {
+            pack_array(
+                arr.clone(),
+                self.num_bits,
+                &mut dst_buffer,
+                &mut dst_idx,
+                &mut dst_offset,
+            )?;
+        }
+
+        let data_type = arrays[0].data_type();
+        Ok((
+            EncodedBuffer {
+                parts: vec![dst_buffer.into()],
+            },
+            EncodedBufferMeta {
+                bits_per_value: (data_type.byte_width() * 8) as u64,
+                bitpacked_bits_per_value: Some(self.num_bits),
+                compression_scheme: None,
+            },
+        ))
+    }
+}
+
+fn pack_array(
+    arr: ArrayRef,
+    num_bits: u64,
+    dst: &mut [u8],
+    dst_idx: &mut usize,
+    dst_offset: &mut u8,
+) -> Result<()> {
+    match arr.data_type() {
+        DataType::UInt8 | DataType::UInt16 | DataType::UInt32 | DataType::UInt64 => {
+            pack_buffers(
+                arr.to_data(),
+                num_bits,
+                arr.data_type().byte_width(),
+                dst,
+                dst_idx,
+                dst_offset,
+            );
+
+            Ok(())
+        }
+        _ => Err(Error::InvalidInput {
+            source: format!("Invalid data type for bitpacking: {}", arr.data_type()).into(),
+            location: location!(),
+        }),
+    }
+}
+
+fn pack_buffers(
+    data: ArrayData,
+    num_bits: u64,
+    byte_len: usize,
+    dst: &mut [u8],
+    dst_idx: &mut usize,
+    dst_offset: &mut u8,
+) {
+    let buffers = data.buffers();
+    for buffer in buffers {
+        pack_bits(buffer, num_bits, byte_len, dst, dst_idx, dst_offset);
+    }
+}
+
+fn pack_bits(
+    src: &[u8],
+    num_bits: u64,
+    byte_len: usize,
+    dst: &mut [u8],
+    dst_idx: &mut usize,
+    dst_offset: &mut u8,
+) {
+    let bit_len = byte_len as u64 * 8;
+
+    let mut mask = 0u64;
+    for _ in 0..num_bits {
+        mask = mask << 1 | 1;
```

Comment:
Minor nit:  Something like `u64::MAX >> (64 - num_bits)` might be more efficient (I might have an off-by-one in that logic)

Reply:
Yes this works, made this change